### ISSUE TITLE
 Different results on groupby data

### ISSUE THREAD

**Issue by nirupamkar** [dmlc/xgboost#3734]
I am using XGB for a classification problem. If I groupby the features and use the count as weight, ideally there should not be any change in the model as loss function should remain same. However I am seeing a change. Please let me know if I am missing something.

np.random.seed(42)
x=pd.DataFrame({"y":[int(np.random.random()>.8) for i in range(100)],
               "x1":[np.round(np.random.random(),1) for i in range(100)],"x2":[np.round(np.random.random(),1) for i in range(100)]})
x["weight"]=1
import xgboost  as xgb

valid_features=["x1","x2"]

xgb_train_data = xgb.DMatrix(x[valid_features], label=x.y.values, 
                                  weight=x.weight.values
                                  )
parameters = {
"boosting_type": 'gbdt', 
"objective" : 'binary:logistic',
    "subsample":.7,
    "random_state":42,
    "num_leaves":64,
    "eval_metric":["logloss"]
}
print "running "
evallist=[(lgb_train_data, 'eval')]
model2 = xgb.train(parameters,
xgb_train_data,10,evallist)

log_loss(x.y,model2.predict(xgb_train_data),sample_weight=x.weight.values)
0.3447492831200361

x1=x.groupby(["x1","x2","y"]).agg({"weight":"sum"}).reset_index()
xgb_train_data = xgb.DMatrix(x1[valid_features], label=x1.y.values, 
                                  weight=x1.weight.values,
                                  )
parameters = {
"boosting_type": 'gbdt', 
"objective" : 'binary:logistic',
    "subsample":.7,
    "seed":42,
    "depth":64,
    "eval_metric":["logloss"]
}
print "running "
evallist=[(lgb_train_data, 'eval')]
model2 = xgb.train(parameters,
xgb_train_data,10,evallist)

log_loss(x1.y,model2.predict(xgb_train_data),sample_weight=x1.weight.values)
0.3524573728069663

1 more interesting observation. If I change subsample=1 the results are matching. So does it indicate the sampling might not be weighted?

**Comment by hcho3** [comment]
Yes, currently sampling due to `subsample` currently does not take account of instance weights.

**Comment by khotilov** [comment]
That reminds me about a git stash that I have made somewhere a while ago when I really needed the subsampling to be weighted... I'll dig it up. It might be more involved now though since we have many updaters.

@nirupamkar even with proper subsampling I would not expect to get the same results (should be rather similar though) since the same values might end up in and out of a subsample.

**Comment by nirupamkar** [comment]
@khotilov I also agree that it should not be same but it should be close enough. 

If there is no weighted subsample, it is giving equal weight to each row for subsample and thereby the subsequent trees will be biased towards a different data distribution. 

So if the weight distribution is skewed the resultant model will have significantly less prediction power compared to a model build with multiple same rows(weight=1). It's actually trying to find a pattern in the less dense areas and not able to predict appropriately in the dense areas.

**Comment by trivialfis** [comment]
Implemented as gradient based sampling.
